### ISSUE TITLE
LXD can't remove volume provided by LVM backend

### ISSUE THREAD

**Issue by syswipe** [lxc/lxd#5806]
* Distribution: Ubuntu
 * Distribution version: 16.04.6 LTS
 * The output of "lxc info":
```
lxc info
config:
  cluster.https_address: 10.0.0.30:8443
  core.https_address: 10.0.0.30:8443
  core.trust_password: true
api_extensions:
- storage_zfs_remove_snapshots
- container_host_shutdown_timeout
- container_stop_priority
- container_syscall_filtering
- auth_pki
- container_last_used_at
- etag
- patch
- usb_devices
- https_allowed_credentials
- image_compression_algorithm
- directory_manipulation
- container_cpu_time
- storage_zfs_use_refquota
- storage_lvm_mount_options
- network
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
- storage_lvm_vg_rename
- storage_lvm_thinpool_rename
- network_vlan
- image_create_aliases
- container_stateless_copy
- container_only_migration
- storage_zfs_clone_copy
- unix_device_rename
- storage_lvm_use_thinpool
- storage_rsync_bwlimit
- network_vxlan_interface
- storage_btrfs_mount_options
- entity_description
- image_force_refresh
- storage_lvm_lv_resizing
- id_map_base
- file_symlinks
- container_push_target
- network_vlan_physical
- storage_images_delete
- container_edit_metadata
- container_snapshot_stateful_migration
- storage_driver_ceph
- storage_ceph_user_name
- resource_limits
- storage_volatile_initial_source
- storage_ceph_force_osd_reuse
- storage_block_filesystem_btrfs
- resources
- kernel_limits
- storage_api_volume_rename
- macaroon_authentication
- network_sriov
- console
- restrict_devlxd
- migration_pre_copy
- infiniband
- maas_network
- devlxd_events
- proxy
- network_dhcp_gateway
- file_get_symlink
- network_leases
- unix_device_hotplug
- storage_api_local_volume_handling
- operation_description
- clustering
- event_lifecycle
- storage_api_remote_volume_handling
- nvidia_runtime
- container_mount_propagation
- container_backup
- devlxd_images
- container_local_cross_pool_handling
- proxy_unix
- proxy_udp
- clustering_join
- proxy_tcp_udp_multi_port_handling
- network_state
- proxy_unix_dac_properties
- container_protection_delete
- unix_priv_drop
- pprof_http
- proxy_haproxy_protocol
- network_hwaddr
- proxy_nat
- network_nat_order
- container_full
- candid_authentication
- backup_compression
- candid_config
- nvidia_runtime_config
- storage_api_volume_snapshots
- storage_unmapped
- projects
- candid_config_key
- network_vxlan_ttl
- container_incremental_copy
- usb_optional_vendorid
- snapshot_scheduling
- container_copy_project
- clustering_server_address
- clustering_image_replication
- container_protection_shift
- snapshot_expiry
- container_backup_override_pool
- snapshot_expiry_creation
- network_leases_location
- resources_cpu_socket
- resources_gpu
- resources_numa
- kernel_features
- id_map_current
- event_location
- storage_api_remote_volume_snapshots
- network_nat_address
- container_nic_routes
- rbac
- cluster_internal_copy
- seccomp_notify
- lxc_features
- container_nic_ipvlan
api_status: stable
api_version: "1.0"
auth: trusted
public: false
auth_methods:
- tls
environment:
  addresses:
  - 10.0.0.30:8443
  architectures:
  - x86_64
  - i686
  certificate: |
    -----BEGIN CERTIFICATE-----
    MIIFUzCCAzugAwIBAgIRAKa2GJpBtESWVBUr/5QftiAwDQYJKoZIhvcNAQELBQAw
    NjEcMBoGA1UEChMTbGludXhjb250YWluZXJzLm9yZzEWMBQGA1UEAwwNcm9vdEBq
    b2JwdWx6MTAeFw0xODA2MjAxNDI4NDlaFw0yODA2MTcxNDI4NDlaMDYxHDAaBgNV
    BAoTE2xpbnV4Y29udGFpbmVycy5vcmcxFjAUBgNVBAMMDXJvb3RAam9icHVsejEw
    ggIiMA0GCSqGSIb3DQEBAQUAA4ICDwAwggIKAoICAQDGH57cSdm2CTCUsoR+N9Y3
    JNWoXFKmcGa4F4/ZM1IyhzK7ZpWesXNPw1llmKoeUIzt6041XwOR3tk/pU95WDha
    ZXD2ilNiP8ws/5mHWTgfy/56SNbLZIJe6/WfIl9E7eCZewYjCg5XVWw6l+FMzFJk
    E11k6cPsbkIiUMBRzi/3OsfgoBTe34YO5yN7cTi9Y/EtSuCiY2Xav8TTRZ0w4jf3
    FjUyUKNe8+2oQQXgZRypE9o3tjKtP8tAUq3IS3beasin+K0s7fHTrCvHJijFc2gV
    s9jMU+e4HM+o97jsgxHpIpJTA9LUe5bJNvCgLEOX9MmkReyZjBqE5GVzrj3hE+he
    ByAoPN+D8x8AOagNFYPWsprXnvlUgVAxYqzBw5mUT2yzo3iQ+ECiyYic+U++KFSt
    9MRjp0k53uxoB23DqtFc/boM1A4/BerLm/CxiBMP5cYUJc5IC9x1yQXU+j/TgwfN
    dVejyFAYXgn/rbUO8xKTzvNQ3UcDp+3c0hmzZTursTdjmUJISGChltxmr3HefJqs
    V6dbi0XBSVbO9wYjj4CUBPqAEZ6zT6uxkb1V9NyXM0i1tsyV3CFBIu/8aHt947yZ
    XLhZDP5A6BUCP+itt1LTf2ETVBc3dh76XSv7M6jFbb1xfjUJ4RFPuX9yyr1wC2xh
    RXgD/rNoTwwi3qrUqwp/awIDAQABo1wwWjAOBgNVHQ8BAf8EBAMCBaAwEwYDVR0l
    BAwwCgYIKwYBBQUHAwEwDAYDVR0TAQH/BAIwADAlBgNVHREEHjAcgghqb2JwdWx6
    MYcEW719IocECgAAFIcECgAACjANBgkqhkiG9w0BAQsFAAOCAgEAV3M3/lM88wwf
    nAoZJfp8y3jJG8lk418aZT6Ki3BL6//I2ldslgkEcXMAPc7ovr+Xy9AhbMvdOc6h
    oiiDG8njrgaoHBp5LascU5b5F9heQo47BpWM+UOB7jtp1/r4O4E6yichB9wElwrs
    M2op9AnG2CcnqNKW3XTbdWY7eEuwbml3mjGuUJBKGV29qwT9G8+q2/PB1R+7SnTm
    f5gISCYkwiPlixGqRQh0kX63djN2SCqNy5lzDJDRQS07qtHULcmrNEHxf/wt7P1Y
    rtIjBqr1sS2qT04VtQxApsKhlBvouhWYJjhrbj6SaiMv1R5zSUOv73VdICHawnN8
    wosM7RlL4f8jVFc/daIWe8SIqjObUbYXDF5y74v0xM8bSS7ZbQYN61J8Tr+Sfhxk
    H4dDHIQEPKNJ0tauGnU4ZCs/8htGjoVpiN8E9T8DTuetfOpJno2np1PIDMGJ1dKs
    GvWZS398O/dV8bgwh5n4PJaz9TeosqqFmYv9fxu3f73oFDyWvN83QSbACU+H1P4p
    9g75ktTUq13SMztyxO2x7BTm4MsTmo4m7AAnH93I+VVJU0aMdnM9TBcDra8r8GC+
    rDwjqPFjso66d/F3iRcXUCgFrK1NEswFcirOT6ts2b6f039gFZ1/Dp6Ync/r6LOJ
    9rcXw9i4lanGNXaHoOnVdOcSoipsf9E=
    -----END CERTIFICATE-----
  certificate_fingerprint: b5e4f4aa59774e57312c6ee25c2060dcfd36db54150c988a5456d351d6b56697
  driver: lxc
  driver_version: 3.1.0
  kernel: Linux
  kernel_architecture: x86_64
  kernel_features:
    netnsid_getifaddrs: "false"
    seccomp_listener: "false"
    shiftfs: "false"
    uevent_injection: "false"
    unpriv_fscaps: "true"
  kernel_version: 4.4.0-141-generic
  lxc_features:
    mount_injection_file: "true"
    network_gateway_device_route: "false"
    network_ipvlan: "false"
    network_l2proxy: "false"
    seccomp_notify: "false"
  project: default
  server: lxd
  server_clustered: true
  server_name: cl-node-3
  server_pid: 11886
  server_version: "3.13"
  storage: lvm
  storage_version: 2.02.133(2) (2015-10-30) / 1.02.110 (2015-10-30) / 4.34.0
```

# Issue description
I can't remove volumes provided by LVM storage backend. Also, I can't move containers between nodes: the process of migration fails because it can't delete source LV. For newly created containers problem doesn't reproduce.

# Steps to reproduce

 1. volume looks like:
```
|         TYPE         |                               NAME                               | DESCRIPTION | USED BY | LOCATION  |
| custom               | demo-storage                                                     |             | 0       | cl-node-2 |
```
 2. i'm trying to delete this volume:
```
lxc storage volume delete local demo-storage 
Error: Could not remove LV named demo--storage
```

**Comment by stgraber** [comment]
Can you look at your LXD log for a more complete error?

**Comment by stgraber** [comment]
The output of `lvs` would also be useful

**Comment by syswipe** [comment]
corresponding events from log:
```
t=2019-05-31T21:13:13+0200 lvl=info msg="Shutting down container" action=shutdown created=2018-11-13T15:15:26+0100 ephemeral=false name=prod-docker-1 project=default timeout=-1s used=2019-01-22T18:04:50+0100
t=2019-05-31T21:13:26+0200 lvl=info msg="Shut down container" action=shutdown created=2018-11-13T15:15:26+0100 ephemeral=false name=prod-docker-1 project=default timeout=-1s used=2019-01-22T18:04:50+0100
t=2019-05-31T21:16:12+0200 lvl=info msg="Deleting container" created=2018-11-13T15:15:26+0100 ephemeral=false name=prod-docker-1 project=default used=2019-01-22T18:04:50+0100
t=2019-05-31T21:16:23+0200 lvl=eror msg="Could not remove LV \"prod--docker--1\":   Logical volume local/containers_prod--docker--1 contains a filesystem in use.\n" 
t=2019-05-31T21:16:23+0200 lvl=eror msg="Failed deleting container storage" err="Could not remove LV named prod--docker--1" name=prod-docker-1
```
lvs
```
  LV                                                                      VG    Attr       LSize   Pool        Origin                                                                  Data%  Meta%  Move Log Cpy%Sync Convert
  LXDThinPool                                                             local twi-aotz-- 891.25g                                                                                     15.36  1.73                            
  containers_artifactory--repo                                            local Vwi-aotz--  10.00g LXDThinPool                                                                         32.08                                  
  containers_demo                                                         local Vwi-aotz-- 100.00g LXDThinPool                                                                         48.52                                  
  containers_mongodb--2                                                   local Vwi-aotz--  10.00g LXDThinPool images_e09ab6e15980dad3405c14d7d8a4d667870724273f22f43899d28c7f56fc45f7 84.04                                  
  containers_pgsql--2                                                     local Vwi-aotz--  10.00g LXDThinPool                                                                         14.01                                  
  containers_prod--2                                                      local Vwi-aotz--  10.00g LXDThinPool                                                                         35.91                                  
  containers_prod--docker--1                                              local Vwi-aotz--  20.00g LXDThinPool images_d9b8b60b834537fa1f4c1163566ffe7ef65ed963b88c06cc28ec7f46c8b18956 31.81                                  
  containers_rmq--2                                                       local Vwi-aotz--  10.00g LXDThinPool                                                                         14.48                                  
  custom_demo-storage                                                     local Vwi-aotz--  50.00g LXDThinPool                                                                         25.80                                  
  custom_demo-storage-copy                                                local Vwi-aotz--  50.00g LXDThinPool custom_prod-storage-2                                                   32.03                                  
  custom_demo-storage-new                                                 local Vwi-aotz--  50.00g LXDThinPool custom_prod-storage-2                                                   29.58                                  
  custom_mongodb-2-data                                                   local Vwi-aotz--  40.00g LXDThinPool                                                                         39.96                                  
  custom_pgsql-2-pgdata-1                                                 local Vwi-aotz--  30.00g LXDThinPool                                                                         80.50                                  
  custom_pgsql-backup                                                     local Vwi-aotz--  20.00g LXDThinPool                                                                         29.81                                  
  custom_prod-storage-2                                                   local Vwi-aotz--  50.00g LXDThinPool                                                                         33.01                                  
  images_9743fc7ecac0ba3fdc04a037b77a62af3472e58755cf2db1feeadad9809e90c5 local Vwi-a-tz--  10.00g LXDThinPool                                                                         8.98                                   
  images_d9b8b60b834537fa1f4c1163566ffe7ef65ed963b88c06cc28ec7f46c8b18956 local Vwi-a-tz--  10.00g LXDThinPool                                                                         9.02                                   
  images_e09ab6e15980dad3405c14d7d8a4d667870724273f22f43899d28c7f56fc45f7 local Vwi-a-tz--  10.00g LXDThinPool                                                                         9.03                                   

```

**Comment by stgraber** [comment]
```
t=2019-05-31T21:16:23+0200 lvl=eror msg="Could not remove LV \"prod--docker--1\":   Logical volume local/containers_prod--docker--1 contains a filesystem in use.\n" 
```

That's going to be a problem for LV removal.

Can you show:

```
grep containers_prod--docker-1 /proc/*/mountinfo
```

**Comment by syswipe** [comment]
`grep containers_prod--docker-1 /proc/*/mountinfo
`
output is empty.

**Comment by stgraber** [comment]
Doh, sorry, should be containers_prod--docker--1 instead

**Comment by syswipe** [comment]
output attached
[grep.log](https://github.com/lxc/lxd/files/3252066/grep.log)

**Comment by stgraber** [comment]
Can you show `ps fauxww` too?

The output above looks like you have a container actively running using that LV.

**Comment by syswipe** [comment]
Now the container which is owner, of this LV is running. But at time of migration, it was stopped
[ps.log](https://github.com/lxc/lxd/files/3255937/ps.log)

**Comment by stgraber** [comment]
Can you confirm this is still going on with LXD 3.14?

I don't remember us fixing anything that would make this problem go away but there were a variety of storage fixes so worth checking that this is still happening.

**Comment by syswipe** [comment]
The problem still exists: i can't remove any LV that was detached from containers.

**Comment by syswipe** [comment]
I think my LVs are blocked by smth. I'll reboot servers this weekends during maintenance and will try to remove LV's and do migration again.

**Comment by kkremitzki** [comment]
I'm not sure if this is the same problem or not, but I arrived here by searching for the same error message which I'm seeing in a fresh Debian 10 droplet on Digital Ocean.

I first installed snap and then using that installed LXD (3.14). After running `lxd init` and creating storage with the LVM backend, I do `lxc launch ubuntu:18.04 foo`. Then, waiting a bit to make sure the new container is not in some intermediate starting state, I do `lxc rm -f foo`. This results in the container stopping, but I get an error, `Error: Could not remove LV named foo`. Running the command again a second time removes the container/volume as expected. Running `lxc stop` and then `lxc rm` also works as expected.

**Comment by stgraber** [comment]
@syswipe did the reboot unblock things for you?

**Comment by syswipe** [comment]
Hi, @stgraber!
planned reboot was delayed, i'm going to do it ASAP. After that i'll inform you.

**Comment by syswipe** [comment]
@stgraber , i've just rebooted one of the nodes affected. Now i can delete volumes that belong to this node. On the other nodes problem still exists...

**Comment by syswipe** [comment]
now, i rebooted all nodes. problem solved. but the reason remains unclear

**Comment by Ne00n** [comment]
Anyone has a solution to this?
Still happening, reboot does solve it but no permanent fix.

The container is just stock Debian 10, nothing running on it.
I did find the issue to be operating system independent.

https://pastebin.com/raw/GQ8rGnXE

**Comment by Ne00n** [comment]
LXD reports the container as stopped, but its still mounted?
| lxcc39dbb31 | STOPPED |                    |                                            | CONTAINER | 0         |

**Comment by tomponline** [comment]
If you unmount each one of those repeated mounts that should solve it:

```
sudo umount /var/snap/lxd/common/lxd/storage-pools/secondary/containers/lxcc39dbb31
```

You may need to do it inside the snap's namespace, so try also:

```
sudo nsenter --mount=/run/snapd/ns/lxd.mnt -- umount /var/snap/lxd/common/lxd/storage-pools/secondary/containers/lxcc39dbb31
```

Repeat until all the mounts for that instance are gone.

**Comment by Ne00n** [comment]
> If you unmount each one of those repeated mounts that should solve it:
> 
> ```
> sudo umount /var/snap/lxd/common/lxd/storage-pools/secondary/containers/lxcc39dbb31
> ```
> 
> You may need to do it inside the snap's namespace, so try also:
> 
> ```
> sudo nsenter --mount=/run/snapd/ns/lxd.mnt -- umount /var/snap/lxd/common/lxd/storage-pools/secondary/containers/lxcc39dbb31
> ```
> 
> Repeat until all the mounts for that instance are gone.

Is there an API for this? any way to automate this?
Because this problem happens randomly.

Any plans to fix this?

**Comment by tomponline** [comment]
We would need a reproducer for it as it does not happen randomly normally (in our test suites or experience).

**Comment by Ne00n** [comment]
> We would need a set of reproducer steps for it to show how it happens as it does not happen randomly normally (in our test suites or experience). As well as details of the host system, kernel version etc.

So, I should ditch LVM? or is this a known bug to other storage backends too?
I can reproduce this bug on several machines, running different ubuntu kernels, apparently the kernel dosen't matter at all.

**Comment by Ne00n** [comment]
Where do I need to look, what info do you need?

**Comment by tomponline** [comment]
> So, I should ditch LVM? or is this a known bug to other storage backends too?

Its not a known bug.

This is why we would need more information.

Please post over at https://discuss.linuxcontainers.org/ with your system details, and steps you perform (after a fresh reboot) to get into this mount state. Then we can go through it together and if necessary reopen this issue or open a new issue.

**Comment by Ne00n** [comment]
> > So, I should ditch LVM? or is this a known bug to other storage backends too?
> 
> Its not a known bug.
> 
> This is why we would need more information.
> 
> Please post over at https://discuss.linuxcontainers.org/ with your system details, and steps you perform (after a fresh reboot) to get into this mount state. Then we can go through it together and if necessary reopen this issue or open a new issue.

That is the problem, I don't know the machine gets into this state.
The machine works without issues, until you try to delete it.

There is no specific software running inside the container.
Hence I asked, if any log file maybe gives what you need, so I asked where shall I need to look.

But I will for sure post later.

**Comment by tomponline** [comment]
Thank you. I suggest you start and stop the container repeatedly until it happens.

The LXD logs are stored in /var/snap/lxd/common/lxd/logs/lxd.log
